rust/module/src/lib.rs

```diff
@@ -0,0 +1,355 @@
+// SPDX-License-Identifier: GPL-2.0
+
+//! Implements the `module!` macro magic
+
+extern crate proc_macro;
+
+use proc_macro::{TokenStream, TokenTree, Group, Delimiter, token_stream};
+
+fn expect_ident(it: &mut token_stream::IntoIter) -> String {
+    if let TokenTree::Ident(ident) = it.next().unwrap() {
+        ident.to_string()
+    } else {
+        panic!("Expected Ident");
+    }
+}
+
+fn expect_punct(it: &mut token_stream::IntoIter) -> char {
+    if let TokenTree::Punct(punct) = it.next().unwrap() {
+        punct.as_char()
+    } else {
+        panic!("Expected Punct");
+    }
+}
+
+fn expect_literal(it: &mut token_stream::IntoIter) -> String {
+    if let TokenTree::Literal(literal) = it.next().unwrap() {
+        literal.to_string()
+    } else {
+        panic!("Expected Literal");
+    }
+}
+
+fn expect_group(it: &mut token_stream::IntoIter) -> Group {
+    if let TokenTree::Group(group) = it.next().unwrap() {
+        group
+    } else {
+        panic!("Expected Group");
+    }
+}
+
+fn expect_end(it: &mut token_stream::IntoIter) {
+    if let None = it.next() {
+    } else {
+        panic!("Expected end");
+    }
+}
+
+fn get_ident(it: &mut token_stream::IntoIter, expected_name: &str) -> String {
+    assert_eq!(expect_ident(it), expected_name);
+    assert_eq!(expect_punct(it), ':');
+    let ident = expect_ident(it);
+    assert_eq!(expect_punct(it), ',');
+    ident
+}
+
+fn get_literal(it: &mut token_stream::IntoIter, expected_name: &str) -> String {
+    assert_eq!(expect_ident(it), expected_name);
+    assert_eq!(expect_punct(it), ':');
+    let literal = expect_literal(it);
+    assert_eq!(expect_punct(it), ',');
+    literal
+}
+
+fn get_group(it: &mut token_stream::IntoIter, expected_name: &str) -> Group {
+    assert_eq!(expect_ident(it), expected_name);
+    assert_eq!(expect_punct(it), ':');
+    let group = expect_group(it);
+    assert_eq!(expect_punct(it), ',');
+    group
+}
+
+fn get_byte_string(it: &mut token_stream::IntoIter, expected_name: &str) -> String {
+    let byte_string = get_literal(it, expected_name);
+
+    assert!(byte_string.starts_with("b\""));
+    assert!(byte_string.ends_with("\""));
+
+    byte_string[2..byte_string.len() - 1].to_string()
+}
+
+fn build_modinfo_builtin_string(module_name: &str, name: &str, string: &str) -> String {
+    format!(
+        "
+            // Built-in modules prefix their modinfo strings by `module_name.`
+            #[cfg(not(MODULE))]
+            #[link_section = \".modinfo\"]
+            #[used]
+            pub static {name}: [u8; {length}] = *b\"{module_name}.{string}\\0\";
+        ",
+        module_name = module_name,
+        name = name,
+        length = module_name.len() + 1 + string.len() + 1,
+        string = string,
+    )
+}
+
+fn build_modinfo_string(module_name: &str, name: &str, string: &str) -> String {
```

Comment:
The code for this and `build_modinfo_builtin_string` are the same. Can they be implemented by having the caller do the `{module_name}.{string}` concat?

Reply:
Yeah, it should be cleaned up -- also the callers can be cleaned up to avoid repeating the `__{}__X` format string which in the end I used in a few places. I'll clean it up.

Reply:
Both done -- callers look way better now.